plugin/context.go

```diff
@@ -2,48 +2,123 @@ package plugin
 
 import (
 	"context"
-	"fmt"
 	"path/filepath"
 
+	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/events"
 	"github.com/containerd/containerd/log"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
 )
 
-// NewContext returns a new plugin InitContext
-func NewContext(ctx context.Context, plugins map[Type]map[string]interface{}, root, state, id string) *InitContext {
-	return &InitContext{
-		plugins: plugins,
-		Root:    filepath.Join(root, id),
-		State:   filepath.Join(state, id),
-		Context: log.WithModule(ctx, id),
-	}
-}
-
 // InitContext is used for plugin inititalization
 type InitContext struct {
+	Context context.Context
 	Root    string
 	State   string
-	Address string
-	Context context.Context
 	Config  interface{}
+	Address string
 	Events  *events.Exchange
 
-	plugins map[Type]map[string]interface{}
+	Meta *Meta // plugins can fill in metadata at init.
+
+	plugins *PluginSet
+}
+
+// NewContext returns a new plugin InitContext
+func NewContext(ctx context.Context, r *Registration, plugins *PluginSet, root, state string) *InitContext {
+	return &InitContext{
+		Context: log.WithModule(ctx, r.URI()),
+		Root:    filepath.Join(root, r.URI()),
+		State:   filepath.Join(state, r.URI()),
+		Meta: &Meta{
+			Exports: map[string]string{},
+		},
+		plugins: plugins,
+	}
 }
 
 // Get returns the first plugin by its type
 func (i *InitContext) Get(t Type) (interface{}, error) {
-	for _, v := range i.plugins[t] {
-		return v, nil
+	return i.plugins.Get(t)
+}
+
+// Meta contains information gathered from the registration and initialization
+// process.
+type Meta struct {
```

Comment:
Meta -> Introspection?

Reply:
Nope, please look at the other data types.

Reply:
I know `Meta` is consistently used across the code, but this name looks too generic to me.

Reply:
This is literally metadata about the plugin. "Introspection" is the name of a service, so doesn't make sense. "Plugin" is already taken. This is a good name for this purpose.

Reply:
it's `plugin.Meta` it works for me (I agree that without considering the package it's confusing, but we do have packages in Go :))